### ISSUE TITLE
wrong keypoint coordinates

### ISSUE THREAD

**Issue by christian-rauch** [MagicLeapResearch/SuperPointPretrainedNetwork#12]
The keypoint coordinates are extracted from the heatmap via:
https://github.com/magicleap/SuperPointPretrainedNetwork/blob/1fda796addba9b6f8e79d586a3699700a86b1cea/demo_superpoint.py#L251-L257

The indices returned by `np.where`, respectively `np.nonzero` are in order of the dimensions. In a NumPy array, the first dimension is along the rows (i.e. the `y` coordinates in an image), the second dimension is along the columns (i.e. the `x` coordinates in an image).

Hence, syntactically this should be:
```Python
ys, xs = np.where(heatmap >= self.conf_thresh) # Confidence threshold.
# [...]
pts[0, :] = xs
pts[1, :] = ys
pts[2, :] = heatmap[ys, xs]
```

The actual variable name does not matter later, as those values are correctly normalised.

**Comment by ddetone** [comment]
Good catch! I think the variable names are wrong, but the output should be correct.

Unfortunately I don't have write privileges and can't update the code :/

**Comment by christian-rauch** [comment]
Yes, the output is not affected since the inverted meaning is inverted again when accessing the heatmap and for the `grid_sample`. 

When porting this to other languages, e.g. C++, you have to sample the feature map with (x,y) tuples. The python code currently suggests that these should be (y,x) tuples, resulting in the wrong descriptors.

**Comment by laisimiao** [comment]
@christian-rauch I try a small example, and I feel no wrong:
```python
import numpy as np
a = np.random.randn(3, 3)
a
Out[4]: 
array([[ 0.24942095,  0.8585088 , -0.1106783 ],
       [ 0.07011566,  0.21370989,  0.81164736],
       [ 2.40070676, -0.11000709, -1.76859847]])
np.where(a > 0)
Out[5]: (array([0, 0, 1, 1, 1, 2]), array([0, 1, 0, 1, 2, 0]))
```
Where do I understand wrongly?
np.where return (x, y)

**Comment by christian-rauch** [comment]
> Where do I understand wrongly?
> np.where return (x, y)

The first row in your matrix `a` is indeed the first column of the image. That means an index in your first matrix row is the index of the first image column, which is 'y'.

You will see this better when you visualise the data:
```Python3
import numpy as np
import matplotlib.pyplot as plt

if __name__ == "__main__":
    a = np.zeros(shape=(5,7))

    # set pixel at x=5 and y=2 to 1
    a[2,5] = 1

    # find the coordinate of value '1'
    y,x = np.where(a==1)
    print("point (x,y): ",x[0],",",y[0])
    plt.imshow(a)
    plt.show()

```

The image is:
![test](https://user-images.githubusercontent.com/8226248/93887452-a2a6c880-fcde-11ea-978d-350105af0c12.png)

And the `1` (yellow square) is located at:
```
point (x,y):  5 , 2
```

**Comment by laisimiao** [comment]
@christian-rauch Oh I got it, thanks for your patient and vivid explanation.
---
BTW, could you provide a script that can draw `Qualitative Results` of a pair of images like paper show(see followiing)? Thank you for your help
![image](https://user-images.githubusercontent.com/37405764/93954256-d554cd80-fd7f-11ea-971b-b6041973e09a.png)

**Comment by christian-rauch** [comment]
@laisimiao I opened pull request #14 to fix this issue. Is this sufficient to demonstrate that it works?

Btw, at https://github.com/magicleap/SuperPointPretrainedNetwork/blob/1fda796addba9b6f8e79d586a3699700a86b1cea/demo_superpoint.py#L206
this was already done correctly.

**Comment by yvaine** [comment]
Hi, I think @christian-rauch  is right, and should fix, because 
`samp_pts[0, :] = (samp_pts[0, :] / (float(W)/2.)) - 1`.
the samp_pts[0] represent x axis
